src/utils.rs

```diff
@@ -156,17 +156,22 @@ pub fn is_char_boundary<const N: usize>(s: &ArrayString<N>, idx: u8) -> Result<(
 #[inline]
 pub(crate) fn truncate_str(slice: &str, size: u8) -> &str {
     trace!("Truncate str: {} at {}", slice, size);
-    if slice.is_char_boundary(size.into()) {
-        unsafe { slice.get_unchecked(..size.into()) }
-    } else if (size as usize) < slice.len() {
-        let mut index = size.saturating_sub(1) as usize;
-        while !slice.is_char_boundary(index) {
-            index = index.saturating_sub(1);
-        }
-        unsafe { slice.get_unchecked(..index) }
-    } else {
-        slice
+    let mut size = size as usize;
```

Comment:
I like this idea that we only need to subtract at most 4 bytes, but I'm not sure if manually unrolling the loop is great. Shouldn't the compiler be able to optimize a simple for?

If not then sure, let's unroll it manually.

Reply:
In my tests the compiler will unroll 7 times which is too much. This version makes it a lot faster according to the benchmarks of push truncate, but we should do a comparative one specific to the function.

Is_char_boundary pretty much boils down to a range check and an and with zero check and micro-op fused short jump.
The only way i see to optimize it further is to operate directly on each byte unchecked. It is guaranteed to return even if we don't bound check on a valid utf8 str.

In theory the whole function could boil down to a branchless "magic" function that:
Get offset to start if < 4
Load unaligned 32 bits of the 4 previous bytes clamped to start
NAnd with magic value
Get trailing zeroes / 4 
Substract that value to the end of the slice clamped by the initial offset

Maybe one branch is more efficient where we do one where n >= 4 and the other where n < 4
For such short jumps that happen once there should be no performance impact from it.
Modern processors preload multiple branches, two should create no problem at all, where a while loop would.

Reply:
I was talking about replacing the while for a `for i in (0..4)` or something like that. But manually unrolling is a great improvement already.

Could you please open another PR, as this one got automatically closed when I renamed master to main?

Also, I would prefer if you could split the changes in the benchmark changes, and the optmizations/making things safer.

I would like to merge the benches, but I need more time to review all of the algo changes.